src/main/java/org/commonjava/cdi/util/weft/ContextSensitiveWeakHashMap.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (C) 2013 Red Hat, Inc. (jdcasey@commonjava.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.cdi.util.weft;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.WeakHashMap;
+import java.util.function.Consumer;
+
+/**
+ * ContextSensitiveWeakHashMap is a subclass of WeakHashMap. But an entry will not be GC-ed until all the threads
+ * ever use it clears their ThreadContext.
+ *
+ * An utility method newSynchronizedContextSensitiveWeakHashMap is provided as a convenient way to obtain a
+ * synchronized ContextSensitiveWeakHashMap instance.
+ *
+ * Created by ruhan on 12/21/17.
+ */
+public class ContextSensitiveWeakHashMap<K, V>
```

Comment:
I'm having a little trouble following the logic of this. It seems like what we need is a variant of HashMap that can count the number of references for a given key, then when the ref count hits zero, removes the key. In Weft, I'd assume this would mean using a UUID per ThreadContext for use in the map (which has a lifecycle that's longer than any single ThreadContext)...then use that UUID to place a referent on each key when the key is added or accessed. Then, when the ThreadContext finalizer runs, it would remove that UUID referent from the key...then removes the key if that results in a zero reference count for that key. Again, this would be working on a map instance that exists **outside** the ThreadContext, but uses UUIDs which are associated with ThreadContexts to do reference counting.

It seems like this class does the inverse. It keeps a UUID for the whole map, and uses that UUID to keep a set of keys in the ThreadContext, removing them from the ThreadContext when the finalizer runs.

---

**OK** I'm leaving this comment here as a sign of respect. This is a really elegant solution...your GC-fu is strong. 😄 

Basically, you're using the Set in the ThreadContext as a way of making the keys strongly reachable, without really centrally managing the reference counts...the counts are implied by the set of contexts that keep strong references. Each one that runs its finalizers relinquishes its strong reference, and eventually this will allow GC to clear the WeakHashMap of those keys.

VERY nice.

Reply:
You got it. Thanks for the review. 
It had been a hard time for me to come up with a good solution. I had thought about it for some days until I appreciated that ThreadContext is a great tool for implementing such "stronger weak-hashmap", and coded it within an hour.